text/0113-secure-script-container.md

```diff
@@ -0,0 +1,183 @@
+
+A syntax like the following could be provided:
+
+```html
+<secure-script extra-globals="ga"></secure-script>
+```
+
+### Known Limitations
+
+* Currently there is an outstanding issue with synthetic shadow in that the Event object patching happens lazily, so we are unable to prevent its patching at the time of creating the secure container, which means that events are still retargeted. [See LWC PR #1569](https://github.com/salesforce/lwc/pull/1569).
+
+
+### Testing
+
+We will want to gather the top 20 common libraries and implement their behavior inside of `secure-script` to ensure that we are prooperly accounting for all these libraries' use cases. This will be our integration test suite. 
+
+### Prototype Code
+
+```js
+import createSecureEnvironment from '@caridy/sjs/lib/browser-realm';
+
+// creating the new sandbox with all global endowments
+const evaluate = createSecureEnvironment(
+    undefined,
+    window,
+);
+
+evaluate(`
+    // This initialization will prevent any of these APIs to be polyfilled
+    // on the blue realm that can affect this sandbox.
+    const pE = Object.preventExtensions;
+    pE(HTMLElement.prototype);
+    pE(Element.prototype);
+    pE(Node.prototype);
+    pE(Event.prototype);
+    pE(Document.prototype);
+    pE(EventTarget.prototype);
+    pE(MutationObserver.prototype);
+    pE(HTMLCollection.prototype);
+    pE(NodeList.prototype);
+    pE(ShadowRoot.prototype);
+    pE(HTMLSlotElement.prototype);
+    pE(Text.prototype);
+`);
+
+class SecureScript extends HTMLElement {
+    connectedCallback() {
+        const src = this.textContent;
+        if (src) {
+            evaluate(src);
+        }
+    }
+}
+
+customElements.define('secure-script', SecureScript);
+```
+
+## Drawbacks
+
+- We are potentially prolonging the life of our synthetic shadow dom polyfill, instead of encouraging community library authors to adopt shadow DOM semantics.
+- The element may end up being anywhere from 5kb - 20kb.
+
+## Alternatives
+
+- **Allow shadow DOM to be toggled at a more granular level.** There is a forthcoming RFC for allowing component authors to decide if they want to opt in to shadow DOM or not. This would mean we can keep top-level elements and layout components in the light DOM, but base components or managed components can choose to safeguard their internals. Until we allow this, `secure-script` serves as a mitigation for application developers looking to add third party libraries to their apps.
+- Do not apply the shadow DOM polyfill at all: this would allow component authors to violate shadow DOM encapsulation which is not an option.
+- If we don't do this, then companies / websites building on our platform will not be able to integrate with common third party libraries.
+
+## Adoption strategy
+
+TBD
+
+## How we teach this
+
+TBD
+
+## Unresolved questions
+
+- Do we want to support all the existing `script` tag attributes like `type`, `async` and `defer`?
```

Comment:
are there any downsides to doing this? I'd prefer to not have to go implement any of what you just outlined and given that this is effectively a privileged script tag I can see authors adding in those attr thinking they'll work.

Reply:
Yeah. I was going to test if we can extend `HTMLScriptElement` and get some of this behavior for free.

Reply:
I think the semantics of the `<secure-script>` will be by definition async, otherwise we will have to generate an equivalent imperative inline version so evaluation is sync.

Reply:
What is the motivation for matching imperative APIs, just programmatic convenience? Also there's nothing stopping us from having both, right?

```typescript
SecureScript.evaluateSync
SecureScript.evaluate
```

Reply:
I do think like @diervo that `<secure-script>` should be async by nature but for different reasons. Preserving the default script semantic doesn't behavior would require to force HTML parsing to stop, and I don't feel that we should be in this business.

I would be supportive of adding the `defer` attribute.

Reply:
Im not sure how we will be able to honor `defer` though

Reply:
Ok, in today's demo, it supports everything supported by the original script tag, in fact, the way it works is that `<virtual-script>` becomes `<script>` with the same set of attributes and content inside the sandbox.

Now, many of the issues about scripts are for script tags parsed and executed as part of the html document, which is NEVER the case for `<virtual-script>` because it is always inserted dynamically inside the sandboxed iframe. In which case, browsers will do a good job no matter what to not block anything else. I don't think we need to complicate this more.

Reply:
Since the elements are dynamically inserted into the document sandbox, the original script semantic is not preserved regarding timing and evaluation order. In the following example, we can ensure that `script1.js` is evaluated before `script2.js`.

```html
<script src="script1.js"></script>
<script src="script2.js"></script>
```

Now in the next example, we can guarantee the evaluation order since script processing is not done by the HTML parser. In some cases, `script2.js` might evaluate before `script1.js`.

```html
<virtual-script src="script1.js"></virtual-script>
<virtual-script src="script2.js"></virtual-script>
```

This is why I think one of the invariant here is that all `virtual-script`s should be considered `async` by nature.

Reply:
I see... this is a very valid point, thanks for raising this issue @pmdartus , let me think more about it.

Reply:
The order is still preserved when multiple script tags are set as async, so forcing async doesn't solve the problem. For example:

```html
<script src="script1.js" async></script>
<script src="script2.js" async></script>
```

`script2` is always going to evaluate only after `script1` has been downloaded and evaluated.

@caridy can we insert each script into a single sandbox instead of having a separate sandbox per `virtual-script` tag?

Reply:
Initially, that was my assumption, that they will be inserted on the same sandbox, we only built one... or a way to identify them somehow to be anchored to one or a handful of sandboxes.

Reply:
> The order is still preserved when multiple script tags are set as async, so forcing async doesn't solve the problem.

@tedconn This is not the case, there is no execution order between async scripts.